### ISSUE TITLE
LanguageModelCriterion() vs XE

### ISSUE THREAD

**Issue by jpanaro** [ruotianluo/self-critical.pytorch#199]
Apologies in advance if this is not the right place.

So the paper this code this based on states that model was pretrained using cross entropy (XE) and then after a certain number of epochs switches over to self critical RL.

My question is that it appears instead of the default pytorch cross entropy, the function LanguageModelCriterion() is used, and its output is the substitute loss value for the pretraining epochs. 
Furthermore, it appears the values produced by LanguageModelCriterion() vary from the values produced by the default XE function so I am curious as to why this design choice was made and if it provides any benefit over XE.

**Comment by ruotianluo** [comment]
It is the cross entropy loss. The reason I created my own, because I need to deal with the mask, where ignore_index doesn't work because both padding and eos are both 0.

**Comment by jpanaro** [comment]
Ah, ok, that makes sense, I am assuming you needed to eliminate padding, but keep eos as you want the network to predict the eos token.

Also, if you don't mind, I have a few more questions slightly less relevant to this project. 

By default train_sample_n is set to 5, is this due to the fact that each image has 5 ground truth captions? 
I ask as I am working with a german sign language dataset with only one ground truth per video and using WER instead of CIDEr and as such the calculations for reward seem to be slightly more brittle with only a single gt to base the result off of.

Lastly, for the RewardCriterion() function, is this essentially the second half of the final gradient expression for the REINFORCE function? i.e this part:
![RL_eq](https://user-images.githubusercontent.com/49861328/83697733-ca5a0f80-a5cd-11ea-934f-5a0e30cc63ae.png)

**Comment by ruotianluo** [comment]
1. Yes, but train_sample_n can be any number.
2. single gt is fine. I have trained captioning model with SCST when only one gt is available.
3. It is calculating (r... - r...)logp, where the gradient is equivalant to the equation you gave.

**Comment by jpanaro** [comment]
That makes a ton of sense, thank you very much for the assistance.
